src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py

```diff
@@ -0,0 +1,420 @@
+    try:
+        helm_command = _get_helm_command()
+    except CLIError:
+        accumulated_errors |= _error_handler(Errors.HELM_COMMAND_ERROR,
+                                             "Obtain helm command",
+                                             ignore_errors,
+                                             "HELM OPERATIONS FAILED")
+        return accumulated_errors
+
+    # Helm check
+    output = getoutput(helm_command + ' version')
+
+    lines = output.splitlines()
+    regex = re.compile(r'SemVer:"([0-9.vb]+)"')
+    client_version = ""
+    server_version = ""
+
+    for line in lines:
+        if line.startswith('Client'):
+            try:
+                client_version = regex.search(line).group(1)
+            except AttributeError:
+                pass
+        elif line.startswith('Server'):
+            try:
+                server_version = regex.search(line).group(1)
+            except AttributeError:
+                pass
+
+    if client_version in ("", None):
+        accumulated_errors |= _error_handler(Errors.HELM_CLIENT_VERSION_ERROR, "Helm client version", ignore_errors)
+    else:
+        print('Helm client version: {}'.format(client_version))
+
+    if server_version in ("", None):
+        accumulated_errors |= _error_handler(Errors.HELM_SERVER_VERSION_ERROR, "Helm server version", ignore_errors)
+    else:
+        print('Helm server version: {}'.format(server_version))
+
+    return accumulated_errors
+
+
+def _check_health_environment(ignore_errors):
+    accumulated_errors = _get_docker_version(ignore_errors)
+    accumulated_errors |= _get_cli_version(ignore_errors)
+    accumulated_errors |= _get_helm_version(ignore_errors)
+
+    return accumulated_errors
+
+## Checks for the connectivity ##
+
+def _get_registry_status(login_server, ignore_errors):
+    accumulated_errors = 0
+
+    # Check DNS lookup
+    import socket
+
+    registry_ip = ""
+
+    try:
+        registry_ip = socket.gethostbyname(login_server)
+    except socket.gaierror:
+        pass
+
+    if registry_ip == "":
+        accumulated_errors |= _error_handler(Errors.CONNECTIVITY_REGISTRY_STATUS_ERROR,
+                                             "DNS lookup to {}".format(login_server),
+                                             ignore_errors,
+                                             "CONNECTIVITY OPERATIONS FAILED")
+    else:
+        print_pass("DNS lookup to {} at IP {}".format(login_server, registry_ip))
+
+    return accumulated_errors
+
+def _get_endpoint_and_token_status(cmd, login_server, ignore_errors):
+    accumulated_errors = 0
+
+    import requests
+    try:
+        from urllib.parse import urlencode, urlparse, urlunparse
+    except ImportError:
+        from urllib import urlencode
+        from urlparse import urlparse, urlunparse
+
+    # Check access to login endpoint
+    url = 'https://' + login_server + '/v2/'
+    challenge = requests.get(url)
```

Comment:
Another thing I was thinking was, should we do a forward check (i.e., starting from basic ones) or a backward check (i.e., starting from challenge, if failed, run more basic ones, etc.).

Reply:
Actually we are starting from basic ones (sequence is first make a DNS lookup to login server, challenge endpoint and then refresh/access token). As those tests are sequential, we only proceed to a test if all before tests succeeded. this line would not throw since it's only called if DNS lookup to login server succeeded

Reply:
Do you think a backward check is more efficient? such as if challenge succeeded with 401, do we still need to check dns, etc.?

Reply:
Considering the usual scenario where the registry login server is correct, I think that backwards check is indeed more efficient (since it makes one less http request), but from the dev side I think it makes more sense to do a forward check. I'm not sure with which one should we proceed though